compiler/resolution/lateConstCheck.cpp

```diff
@@ -313,4 +313,24 @@ void lateConstCheck(std::map<BaseAST*, BaseAST*> * reasonNotConst)
     // For now, don't check primitives. Compiler can be loose
     // with const-ness on its own internal temporaries.
   }
+
+  // Additionally check that promotion wrappers using a scalar this
+  // don't take it in by `ref`
+  forv_Vec(FnSymbol, fn, gFnSymbols) {
+    if (fn->hasFlag(FLAG_PROMOTION_WRAPPER)) {
+      for_formals(formal, fn) {
+        if (formal->intent == INTENT_REF) {
+          Type* vt = formal->getValType();
+          if (vt->scalarPromotionType == NULL &&
+              !(isAtomicType(vt) || isSyncType(vt) || isSingleType(vt)) &&
+              !formal->hasFlag(FLAG_ERROR_VARIABLE)) {
+            const char* err = "Racy promotion of scalar value";
```

Comment:
It looks like this message doesn't get triggered by your tests

Reply:
yes that's right. Instead we get errors earlier in compilation. I wanted to leave it here so we aren't surpised if the shadow variables don't trigger the error.

Reply:
Works for me!  This is good to go, then :)